### ISSUE TITLE
Ideas for SDL 3

### ISSUE THREAD

**Issue by SDLBugzilla** [libsdl-org/SDL#3519]
# This bug report was migrated from our old Bugzilla tracker.

**Reported in version:** don't know
**Reported for operating system, platform:** All, All

# Comments on the original bug report:

On 2020-01-22 15:55:25 +0000, David Demelier wrote:

> Hi,
> 
> I already sent some ideas by email but I've been convinced by an individual on the IRC channel to do this on the bugzilla to get reviews.
> 
> Here's a list of things I'd love to see and implement (as a developer myself) for SDL 3.
> 
> Any feedback is welcomed.
> 
> # Stricter coding, naming style.
> 
> For the moment there are high number of style issues in the API. Some symbols are SDL_lowerCased and some are SDL_UpperCased. Some are not even prefixed by SDL_.
> 
> Every functions that are tighed to a specific structure should:
> 
> - Be prefixed by its name: SDL_Foo (object), SDL_FooOpen, SDL_FooDestroy, SDL_FooUpdate, etc.
> - Take the object as first argument: SDL_FooUpdate(SDL_Foo *, ...);
> 
> Getter and setters should be named as following:
> 
> - SDL_Namespace(Get|Set)What (e.g. SDL_Renderer(Get|Set)DrawColor)
> 
> Constructors and destructors should be named as following if applicable:
> 
> - SDL_NamespaceOpen (e.g. SDL_JoystickOpen)
> - SDL_NamespaceDestroy (e.g. SDL_JoystickDestroy)
> 
> If objects are not meant to be dynamically allocated, the names "Init" and "Finish" or "Close" can also be used.
> 
> 
> 
> # Proper semantic versioning.
> 
> One SDL 2.0.x version added sensor support, while it should be bugfix instead. The current versioning scheme makes too difficult to check what features were provided and when at a glance.
> 
> What did 2.0.1 provided? What 2.0.2 provided?
> 
> In the head of packagers, they feel like they are bugfixes while they've added new APIs!
> 
> The Semantic Versioning is the most appropriate versioning scheme for libraries.
> 
> - Major : massive breakage allowed in the API (not every year though, somewhat like every 5-6 years).
> - Minor : improvements and new features added with API compatibility honored
> - Patch : simple bugfixes
> 
> 
> 
> # Simplify development process
> 
> At the moment there is no much information about the way to contribute to SDL. One can ping maintainer using patches via mail but are sometimes missed. Also they are linked to an individual and can not be reviewed by others.
> 
> Since SDL does not have a system for reviewing yet, I propose that a development mailing-list is re-opened and people could send their patch via `hg email` as the bugzilla is more tailored for bug reporting.
> 
> 
> 
> # Removing useless platforms.
> 
> - arts: no longer used
> - winrt: too, AFAIK
> 
> 
> 
> # Fixing inconsistencies where some functions use plain primitives types rather than structs.
> 
> If we have structs, then functions should always operate on them. Since C99, compound literals makes life easier while passing temporaries to functions.
> 
> - SDL_SetRenderDrawColor takes 4 ints
> - SDL_SetPaletteColor take a SDL_Color object
> 
> - SDL_RenderDrawLine takes 4 ints
> - SDL_RenderSetClipRect takes a SDL_Rect
> 
> The color handling is sometimes in 4 ints, sometimes SDL_Color.
> 
> 
> 
> # Getting rid of dozens of buildsystems to unique one: CMake.
> 
> There are just too many build systems in the tree and that is too complicated for users, developers and packagers. Also, having only one for a portable library that SDL is, makes difficult to merge all checks/operations in all build systems.
> 
> Also, lots of people do not rely on the SDL CMake configuration package files provided as autoconf does not install them. Having CMake only will simplify the integration and maintainance.
> 
> 
> 
> # Use doxygen documentation to avoid outdated information on the wiki.
> 
> I've literally spent 3 hours to document the sensor API on the wiki. That's a big waste of time. SDL headers are already almost documented in doxygen, it should be completed with missing bits and the only source of truth.
> 
> Changing something in one enum results in editing the wiki as well.
> 
> 
> 
> # Rename pkg-config files
> 
> The SDL pkg-config file is sdl2 while addons are SDL2_*.
> 
> 
> 
> # Making addons first class citizen and built from the same source tree with proper naming.
> 
> The SDL addons feel they are separate projects as they still do not have any public visibility from the libsdl.org website. At a glance, a newcomer don't even know these libraries exist.
> 
> They have the same troubles as the above paragraphs (cmake, consistency, etc.).
> 
> I propose the simple inclusion of these addons with a new naming scheme for them and a simpler API. Also, they will be available as separate shared/static libraries controlled at build time. Thus, people who don't need SDL_mixer or SDL_ttf can simply not build/link theme.
> 
> 
> 
> ## SDL_ttf new API
> 
> Every functions start with SDL_Font.
> 
> - SDL_Font (struct)
> - SDL_FontStyle (enum SDL_FONT_SOLID, SDL_FONT_SHADED, SDL_FONT_BLENDED)
> - TTF_Init -> SDL_FontInit()
> - TTF_OpenFont* -> SDL_FontOpen*
> - TTF_(Get|Set)Font* -> SDL_Font(Get|Set)*
> - TTF_FontHeight -> SDL_FontGetHeight
> - TTF_FontAscent -> SDL_FontGetAscent
> - TTF_FontDescent -> SDL_FontGetDescent
> - TTF_FontLineSkip -> SDL_FontGetLineSkip
> - TTF_FontFaces -> SDL_FontGetFaces
> - TTF_GlyphIsProvided -> SDL_FontIsGlyphProvided
> - TTF_GlyphMetrics -> SDL_FontGetGlyphMetrics
> - TTF_SizeText -> Should be removed, UTF-8 is the standard
> - TTF_SizeUTF8 -> SDL_FontTextMetrics
> - TTF_SizeUNICODE -> Should be removed, UTF-16 is deprecated
> - TTF_Render* -> SDL_FontRender(const char *, SDL_FontStyle) (UTF-8 only)
> 
> 
> 
> ## SDL_image new API
> 
> Every function start with SDL_Image.
> 
> - SDL_ImageFormat (enum SDL_IMAGE_(CUR|ICO|BMP|PNM|...))
> - IMG_Init -> SDL_ImageInit()
> - IMG_Load -> SDL_ImageOpen(const char *path)
> - IMG_Load_RW -> SDL_ImageOpen_RW
> - IMG_LoadTyped_RW -> SDL_ImageOpenTyped_RW(SDL_RWOps *, SDL_ImageFormat)
> - IMG_is* -> SDL_ImageGetType (returns SDL_ImageFormat)
> 
> 
> 
> ## SDL_mixer new API
> 
> Basic functions start with SDL_Mixer.
> 
> - Mix_Init -> SDL_MixerInit
> - Mix_OpenAudio -> SDL_MixerOpen
> - Mix_CloseAudio -> SDL_MixerClose
> - Mix_GetNumChunkDecoders -> SDL_MixerGetNumChunkDecoders
> - Mix_GetChunkDecoder -> SDL_MixerGetChunkDecoder
> 
> Channel and groups functions
> 
> - Mix_AllocateChannels -> SDL_ChannelsAllocate
> - Mix_ReserveChannels -> SDL_ChannelsReserve
> - Mix_Group* -> SDL_ChannelGroup*
> - Mix_HaltGroup -> SDL_ChannelGroupHalt
> - Mix_Volume -> SDL_ChannelVolume
> - Mix_PlayChannel -> SDL_ChannelPlay
> - Mix_PlayChannelTimed -> SDL_ChannelPlayTimed
> - Mix_FadeInChannel -> SDL_ChannelFadeIn
> - Mix_FadeInChannelTimed -> SDL_ChannelFadeInTimed
> - Mix_Pause -> SDL_ChannelPause
> - Mix_Resume -> SDL_ChannelResume
> - Mix_HaltChannel -> SDL_ChannelHalt
> - Mix_ExpireChannel -> SDL_ChannelExpire
> - Mix_FadeOutChannel -> SDL_ChannelFadeOut
> - Mix_ChannelFinished -> SDL_ChannelHasFinished
> - Mix_Playing -> SDL_ChannelIsPlaying
> - Mix_Paused -> SDL_ChannelIsPaused
> - Mix_FadingChannel -> SDL_ChannelGetFading
> - Mix_GetChunk -> SDL_ChannelGetChunk
> 
> Chunk functions
> 
> - Mix_LoadWAV -> SDL_ChunkOpen
> - Mix_LoadWAV_RW -> SDL_ChunkOpen_RW
> - Mix_QuickLoad -> SDL_ChunkQuickOpen
> - Mix_QuickLoad_RW -> SDL_ChunkQuickOpen_RW
> - Mix_FreeChunk -> SDL_ChunkDestroy
> - Mix_VolumeChunk -> SDL_ChunkGetVolume
> 
> Music functions
> 
> - Mix_GetNumMusicDecoders -> SDL_MusicGetNumDecoders
> - Mix_GetMusicDecoder -> SDL_MusicGetDecoder
> - Mix_LoadMUS -> SDL_MusicOpen
> - Mix_FreeMusic -> SDL_MusicDestroy
> - Mix_PlayMusic -> SDL_MusicPlay
> - Mix_FadeInMusic -> SDL_MusicFadeIn
> - Mix_FadeInMusicPos -> SDL_MusicFadeInPos
> - Mix_HookMusic -> SDL_MusicHook
> - Mix_VolumeMusic -> SDL_MusicVolume
> - Mix_PauseMusic -> SDL_MusicPause
> - Mix_ResumeMusic -> SDL_MusicResume
> - Mix_RewindMusic -> SDL_MusicRewind
> - Mix_SetMusicPosition -> SDL_MusicSetPosition
> - Mix_SetMusicCMD -> SDL_MusicSetCommand
> - Mix_HaltMusic -> SDL_MusicHalt
> - Mix_FadeOutMusic -> SDL_MusicFadeOut
> - Mix_HookMusicFinished -> SDL_MusicSetHookFinished
> - Mix_GetMusicType -> SDL_MusicGetType
> - Mix_PlayingMusic -> SDL_MusicIsPlaying
> - Mix_PausedMusic -> SDL_MusicIsPaused
> - Mix_FadingMusic -> SDL_MusicIsFading
> - Mix_GetMusicHookData -> SDL_MusicGetHookFinished
> 
> Effect functions
> 
> - Mix_RegisterEffect -> SDL_EffectRegister
> - Mix_UnregisterEffect -> SDL_EffectUnregister
> - Mix_UnregisterAllEffects -> SDL_EffectUnregisterAll
> - Mix_SetPostMix -> SDL_EffectSetPostMix
> - Mix_SetPanning -> SDL_EffectSetPanning
> - Mix_SetDistance -> SDL_EffectSetDistance
> - Mix_SetPosition -> SDL_EffectSetPosition
> - Mix_SetReverseStereo -> SDL_EffectSetReverseStereo
> 
> 
> 
> ## SDL_net
> 
> To my humble opinion, this library should be removed as they are many networking library in C including POSIX.

On 2020-01-23 02:56:03 +0000, Sam Lantinga wrote:

> I agree with most of this feedback, and like the satellite library naming suggestions, thanks!
> 
> Some comments on the build system:
> 
> Not all platforms have CMake, and many developers don't have CMake installed. Windows developers should be able to use Visual Studio, Mac developers should be able to use Xcode. I should be able to use autoconf. :)
> 
> I'm fine with CMake being the recommended build environment, and even stripping out support for non-UNIX platforms from it, but I've run into enough problems with it over the years that I don't use CMake. Also, note that the official SDL windows binaries are built on Mac using mingw64, to integrate with build scripts and avoid C runtime dependencies, and that currently requires autotools.

On 2020-03-23 22:53:39 +0000,  wrote:

> > # Simplify development process
> 
> I think there should be more discussion on just moving the center of development to github / git. I'm aware that this has been proposed by many people over the years, but I feel it hasn't gotten the proper attention it needs.
> 
> Github or at least git, would make it much easier for people to contribute to the development of SDL.

On 2020-03-24 01:04:30 +0000, Ryan C. Gordon wrote:

> (In reply to daegon.dhsk from comment # 2)
> > Github or at least git, would make it much easier for people to contribute
> > to the development of SDL.
> 
> This is a hard no from me, both to GitHub and git itself.
> 
> --ryan.

On 2020-03-24 01:15:27 +0000,  wrote:

> (In reply to Ryan C. Gordon from comment # 3)
> > (In reply to daegon.dhsk from comment # 2)
> > > Github or at least git, would make it much easier for people to contribute
> > > to the development of SDL.
> > 
> > This is a hard no from me, both to GitHub and git itself.
> > 
> > --ryan.
> 
> Could we get some details on the background?

On 2020-03-24 03:32:52 +0000, Ryan C. Gordon wrote:

> (In reply to daegon.dhsk from comment # 4)
> 
> > Could we get some details on the background?
> 
> Even if I didn't personally think git is a lousy system, Mercurial is working fine for us and changing it just introduces unnecessary friction.
> 
> As for GitHub: we have been burned by cloud providers several times over the past 20 years, and have spent significant effort engineering our infrastructure so we own all of it. Becoming reliant on an unnecessary for-profit company for the most crucial piece of our workflow is not something we're willing to do.
> 
> --ryan.

On 2020-03-24 08:15:06 +0000, David Demelier wrote:

> (In reply to daegon.dhsk from comment # 2)
> > > # Simplify development process
> > 
> > I think there should be more discussion on just moving the center of
> > development to github / git. I'm aware that this has been proposed by many
> > people over the years, but I feel it hasn't gotten the proper attention it
> > needs.
> > 
> > Github or at least git, would make it much easier for people to contribute
> > to the development of SDL.
> 
> I can't understand this hype to request people to move to Git. Git is by far much more complicated than Mercurial so I'd like to know what's “easier” for you to contribute.
> 
> If by easier you mean that you can send a pull request, please not that sending a patch is by far much easier:
> 
>     hg clone http://hg.libsdl.org/SDL && cd SDL
>     vim
>     hg ci
>     hg export OR hg email
> 
> That's it.
> 
> You can't come to a project and ask them to change their tooling, this is a subjective material. I don't like Git but I don't ask projects to move from Git to Mercurial. I don't like meson and I don't request projects that use meson to switch to CMake. I can go for a while.
> 
> The only way we could improve the contributing process for SDL is to provide a detailed documentation because for now there are not that much. So some people send emails and some tasks on bugzilla. We need a unified, detailed process. That's it.

On 2020-04-02 22:25:09 +0000, David Ludwig wrote:

> (In reply to David Demelier from comment # 6)
> > If by easier you mean that you can send a pull request, please not that
> > sending a patch is by far much easier:
> > 
> >     hg clone http://hg.libsdl.org/SDL && cd SDL
> >     vim
> >     hg ci
> >     hg export OR hg email
> > 
> > That's it.
> 
> Wait, vim is now part of the easy route?!  (and I like vim, for some things at least)  ;-)
> 
> In all seriousness, a move to git sounds ok to me, though not one to GitHub and its partially-closed nature.  I find git's overall ecosystem to be *MUCH* nicer than Mercurial's (enough that I am willing to argue that Mercurial isn't always the "easiest" option, not any more at least), and would probably welcome a move to git repos.  I have, within the past few years, lost Mercurial support in some valuable-to-me tools (BitBucket, especially), and finding good, alternative tooling hasn't always been easy.
> 
> 
> As for the initial suggestions, I'm in favor of a lot of it!  I especially like the idea of iterating on SDL's function names.  Having a scheme along the lines of SDL_<subsystem><action> seems very helpful with regards to editor auto-completion.
> 
> I'd like to add one suggestion with regards to naming: include the major version number in the prefix.  I.e. SDL3.h, SDL3_RenderSetDrawColor, etc.  It would make SDL 3.x detection be a matter of finding a header, rather than finding a header and parsing it, which might not always be possible (with C++'s __has_include, for example).  It would also help make it possible to distinguish SDL 1.x and 2.x code from 3.x code, which I could see helping if and when porting stuff to 3.x.
> 
> Regarding dropping WinRT, it's my understanding that Xbox One still requires using those APIs.  Unless that changes, I think removing that support would be bad (though, that platform's SDL code could definitely use some improvement, with at least some pruning of older, WinRT SDK support).
> 
> Regarding build-systems, I like the idea of better CMake integration, but I would like to see file-globbing of SDL's src directory be an option, too, at least for platforms where this is possible.  (One of these days, I should finish my support for this.  I've had prototypes working on Apple-platforms, as well as Windows, at one point or another). 
> 
> As for including satellite libraries in SDL-proper, this makes me leery as it seems like it would introduce a lot of additional dependencies in SDL, which is already inherently complex due to the number of platforms it supports.  Having naming changes within them does sound nice, and perhaps possible without needing to include them in SDL itself.

On 2020-05-26 07:50:59 +0000, Jan Niklas Hasse wrote:

> > Windows developers should be able to use Visual Studio, Mac developers should be able to use Xcode.
> 
> Both can be generated using CMake :)

On 2020-11-28 11:36:22 +0000, Dominik Reichardt wrote:

> > > Windows developers should be able to use Visual Studio, Mac developers should be able to use Xcode.
> > 
> > Both can be generated using CMake :)
> 
> I have only tried a few times but those projects all didn't generate a working xcode project. No idea if this is a problem with their CMake files but if you have to work extensively on the Xcode project file each time you generate it from code, then it is no good, IMO worse than a partially working included one.
> 
> So, when I see there is only CMake, I try to find whether someone else already built an Xcode project file or just turn away and don't bother.
> 
> Generation of the unix stuff generally works though.

On 2020-11-28 11:39:12 +0000, Jan Niklas Hasse wrote:

> > I have only tried a few times but those projects all didn't generate a working xcode project.
> 
> What do you mean by "those projects"?

On 2020-11-28 11:44:23 +0000, Dominik Reichardt wrote:

> > > I have only tried a few times but those projects all didn't generate a working xcode project.
> > 
> > What do you mean by "those projects"?
> 
> the projects that had only Cmake and I tried to generate the Xcode project with

On 2020-11-28 11:59:54 +0000, Jan Niklas Hasse wrote:

> Ah I see. I think some projects don't take Xcode into account and use some unidiomatic CMake constructs that break it.
> 
> It is definitely possible though and works great in my projects. It's a good idea to add a CI job which generates and builds with `-GXcode`.

**Comment by markand** [comment]
Since I was the author of the proposal on the bugzilla but not able to edit this one, let's summarize and remove irrelevant stuff (the hard no to move to git is no longer valid).

---

# Stricter coding, naming style.

For the moment there are high number of style issues in the API. Some symbols are `SDL_lowerCased` and some are `SDL_UpperCased`. Some are not even prefixed by SDL_.

Every functions that are tighed to a specific structure should:

Be prefixed by its name: `SDL_Foo (object)`, `SDL_FooOpen`, `SDL_FooDestroy`, `SDL_FooUpdate`, etc.
Take the object as first argument: `SDL_FooUpdate(SDL_Foo *, ...)`;

Getter and setters should be named as following:

- `SDL_Namespace(Get|Set)What` (e.g. `SDL_Renderer(Get|Set)DrawColor`)

Constructors and destructors should be named as following if applicable:

- `SDL_NamespaceOpen` (e.g. `SDL_JoystickOpen`)
- `SDL_NamespaceDestroy` (e.g. `SDL_JoystickDestroy`)

If objects are not meant to be dynamically allocated, the names "Init" and "Finish" or "Close" can also be used.

# Proper semantic versioning.

SDL 2.0.9 version added sensor support, while it should be bugfix instead. The current versioning scheme makes too difficult to check what features were provided and when at a glance.

What did 2.0.1 provided? What 2.0.2 provided?

In the head of packagers, they feel like they are bugfixes while they've added new APIs!

The [Semantic Versioning](http://semver.org) is the most appropriate versioning scheme for libraries.

- **Major**: massive breakage allowed in the API (not every year though, somewhat like every 5-6 years).
- **Minor**: improvements and new features added with API compatibility honored
- **Patch**: simple bugfixes

# Removing useless platforms.

- *arts:* no longer used
- *winrt*: too, AFAIK

# Fixing inconsistencies

If we have structs, then functions should always operate on them. Since C99, compound literals makes life easier while passing temporaries to functions.

- `SDL_SetRenderDrawColor` takes 4 ints
- `SDL_SetPaletteColor` take a SDL_Color object
- `SDL_RenderDrawLine` takes 4 ints
- `SDL_RenderSetClipRect` takes a SDL_Rect

The color handling is sometimes in 4 ints, sometimes SDL_Color.

# Getting rid of dozens of buildsystems to unique one: CMake.

There are just too many build systems in the tree and that is too complicated for users, developers and packagers. Also, having only one for a portable library that SDL is, makes difficult to merge all checks/operations in all build systems.

Also, lots of people do not rely on the SDL CMake configuration package files provided as autoconf does not install them. Having CMake only will simplify the integration and maintainance.

# Use doxygen documentation to avoid outdated information on the wiki

I've literally spent 3 hours to document the sensor API on the wiki. That's a big waste of time. SDL headers are already almost documented in doxygen, it should be completed with missing bits and the only source of truth.

Changing something in one enum results in editing the wiki as well.

# Rename pkg-config files

The SDL pkg-config file is `sdl2` while addons are `SDL2_*`.

# Making addons first class citizen and built from the same source tree with proper naming

The SDL addons feel they are separate projects as they still do not have any public visibility from the libsdl.org website. At a glance, a newcomer don't even know these libraries exist.

They have the same troubles as the above paragraphs (cmake, consistency, etc.).

I propose the simple inclusion of these addons with a new naming scheme for them and a simpler API. Also, they will be available as separate shared/static libraries controlled at build time. Thus, people who don't need SDL_mixer or SDL_ttf can simply not build/link theme.

## SDL_ttf new API

Every functions start with SDL_Font.

- SDL_Font (struct)
- SDL_FontStyle (enum SDL_FONT_SOLID, SDL_FONT_SHADED, SDL_FONT_BLENDED)
- TTF_Init -> SDL_FontInit()
- TTF_OpenFont* -> SDL_FontOpen*
- TTF_(Get|Set)Font* -> SDL_Font(Get|Set)*
- TTF_FontHeight -> SDL_FontGetHeight
- TTF_FontAscent -> SDL_FontGetAscent
- TTF_FontDescent -> SDL_FontGetDescent
- TTF_FontLineSkip -> SDL_FontGetLineSkip
- TTF_FontFaces -> SDL_FontGetFaces
- TTF_GlyphIsProvided -> SDL_FontIsGlyphProvided
- TTF_GlyphMetrics -> SDL_FontGetGlyphMetrics
- TTF_SizeText -> Should be removed, UTF-8 is the standard
- TTF_SizeUTF8 -> SDL_FontTextMetrics
- TTF_SizeUNICODE -> Should be removed, UTF-16 is deprecated
- TTF_Render* -> SDL_FontRender(const char *, SDL_FontStyle) (UTF-8 only)

## SDL_image new API

Every function start with SDL_Image.

- SDL_ImageFormat (enum SDL_IMAGE_(CUR|ICO|BMP|PNM|...))
- IMG_Init -> SDL_ImageInit()
- IMG_Load -> SDL_ImageOpen(const char *path)
- IMG_Load_RW -> SDL_ImageOpen_RW
- IMG_LoadTyped_RW -> SDL_ImageOpenTyped_RW(SDL_RWOps *, SDL_ImageFormat)
- IMG_is* -> SDL_ImageGetType (returns SDL_ImageFormat)

## SDL_mixer new API

Basic functions start with SDL_Mixer.

- Mix_Init -> SDL_MixerInit
- Mix_OpenAudio -> SDL_MixerOpen
- Mix_CloseAudio -> SDL_MixerClose
- Mix_GetNumChunkDecoders -> SDL_MixerGetNumChunkDecoders
- Mix_GetChunkDecoder -> SDL_MixerGetChunkDecoder

Channel and groups functions

- Mix_AllocateChannels -> SDL_ChannelsAllocate
- Mix_ReserveChannels -> SDL_ChannelsReserve
- Mix_Group* -> SDL_ChannelGroup*
- Mix_HaltGroup -> SDL_ChannelGroupHalt
- Mix_Volume -> SDL_ChannelVolume
- Mix_PlayChannel -> SDL_ChannelPlay
- Mix_PlayChannelTimed -> SDL_ChannelPlayTimed
- Mix_FadeInChannel -> SDL_ChannelFadeIn
- Mix_FadeInChannelTimed -> SDL_ChannelFadeInTimed
- Mix_Pause -> SDL_ChannelPause
- Mix_Resume -> SDL_ChannelResume
- Mix_HaltChannel -> SDL_ChannelHalt
- Mix_ExpireChannel -> SDL_ChannelExpire
- Mix_FadeOutChannel -> SDL_ChannelFadeOut
- Mix_ChannelFinished -> SDL_ChannelHasFinished
- Mix_Playing -> SDL_ChannelIsPlaying
- Mix_Paused -> SDL_ChannelIsPaused
- Mix_FadingChannel -> SDL_ChannelGetFading
- Mix_GetChunk -> SDL_ChannelGetChunk

Chunk functions

- Mix_LoadWAV -> SDL_ChunkOpen
- Mix_LoadWAV_RW -> SDL_ChunkOpen_RW
- Mix_QuickLoad -> SDL_ChunkQuickOpen
- Mix_QuickLoad_RW -> SDL_ChunkQuickOpen_RW
- Mix_FreeChunk -> SDL_ChunkDestroy
- Mix_VolumeChunk -> SDL_ChunkGetVolume

Music functions

- Mix_GetNumMusicDecoders -> SDL_MusicGetNumDecoders
- Mix_GetMusicDecoder -> SDL_MusicGetDecoder
- Mix_LoadMUS -> SDL_MusicOpen
- Mix_FreeMusic -> SDL_MusicDestroy
- Mix_PlayMusic -> SDL_MusicPlay
- Mix_FadeInMusic -> SDL_MusicFadeIn
- Mix_FadeInMusicPos -> SDL_MusicFadeInPos
- Mix_HookMusic -> SDL_MusicHook
- Mix_VolumeMusic -> SDL_MusicVolume
- Mix_PauseMusic -> SDL_MusicPause
- Mix_ResumeMusic -> SDL_MusicResume
- Mix_RewindMusic -> SDL_MusicRewind
- Mix_SetMusicPosition -> SDL_MusicSetPosition
- Mix_SetMusicCMD -> SDL_MusicSetCommand
- Mix_HaltMusic -> SDL_MusicHalt
- Mix_FadeOutMusic -> SDL_MusicFadeOut
- Mix_HookMusicFinished -> SDL_MusicSetHookFinished
- Mix_GetMusicType -> SDL_MusicGetType
- Mix_PlayingMusic -> SDL_MusicIsPlaying
- Mix_PausedMusic -> SDL_MusicIsPaused
- Mix_FadingMusic -> SDL_MusicIsFading
- Mix_GetMusicHookData -> SDL_MusicGetHookFinished

Effect functions

- Mix_RegisterEffect -> SDL_EffectRegister
- Mix_UnregisterEffect -> SDL_EffectUnregister
- Mix_UnregisterAllEffects -> SDL_EffectUnregisterAll
- Mix_SetPostMix -> SDL_EffectSetPostMix
- Mix_SetPanning -> SDL_EffectSetPanning
- Mix_SetDistance -> SDL_EffectSetDistance
- Mix_SetPosition -> SDL_EffectSetPosition
- Mix_SetReverseStereo -> SDL_EffectSetReverseStereo

## SDL_net

To my humble opinion, this library should be removed as they are many networking library in C including POSIX.

**Comment by slouken** [comment]
Here are my thoughts...

# Stricter coding, naming style.

Generally agreed. We'll need to create an sdl2-compat library similar to sdl12-compat that preserves the old ABI for older programs while continuing new development in sdl3 with a more consistent API.

# Proper semantic versioning.

We'd have to look at how library versioning is done on Linux. We need to guarantee that older binaries continue to run with newer versions of SDL, and the current versioning system does that. The version is actually generated by using the scheme in autotools:
```
# Making releases:
# Edit include/SDL_version.h and change the version, then:
#   SDL_MICRO_VERSION += 1;
#   SDL_INTERFACE_AGE += 1;
#   SDL_BINARY_AGE += 1;
# if any functions have been added, set SDL_INTERFACE_AGE to 0.
# if backwards compatibility has been broken,
# set SDL_BINARY_AGE and SDL_INTERFACE_AGE to 0.
```
and it turns out that this guarantees binary compatibility in the way that we need.

In terms of knowing when functions have been added, we now include that information in the headers and wiki so it's easy to find.

# Removing useless platforms.

WinRT is still used by the UWP code for Windows Store games and Xbox, so it will continue to be relevant.

# Fixing inconsistencies

> If we have structs, then functions should always operate on them. Since C99, compound literals makes life easier while passing temporaries to functions.

We should review what compilers and platforms are still supported. I'm still occasionally surprised that people are still using platforms that require C89 code. But I agree we should be consistent here, with an eye toward the most efficient and easy to use interfaces.

# Getting rid of dozens of buildsystems to unique one: CMake.

I think we should bite the bullet and just pick CMake over autotools. Maintaining two separate configuration and build systems is too high a tax, and enough people (and platform maintainers) are using CMake that it seems like the right choice for SDL 3. That said, we should still keep the Visual Studio and Xcode projects for people who are more comfortable using the IDE's.

# Use doxygen documentation to avoid outdated information on the wiki

This is already done. Ryan has created a bridge between the headers and wiki so each side is automatically updated by changes in the other side.

# Rename pkg-config files

👍 

# Making addons first class citizen and built from the same source tree with proper naming

The problem here, and the reason why the SDL addons have lagged behind in terms of releases, is these depend on a lot of third party libraries that are time consuming to build and maintain in-tree. git submodules may make this easier though, so this is definitely worth reviewing. The ability to easily add modules to your project with CMake makes this much more viable in the future.

I generally agree with the naming suggestions for the third party APIs above.

## SDL_net

Agreed. There are lots more tools out there now than there were when this library was created.

**Comment by markand** [comment]
> # Proper semantic versioning.
> We'd have to look at how library versioning is done on Linux. We need to guarantee that older binaries continue to run with newer versions of SDL, and the current versioning system does that. The version is actually generated by using the scheme in autotools:

Well semantic versioning does not really care about ABI, it only cares about API in the sense that a program that uses a specific minor version may get new features but with never incompatibilities.

> # Fixing inconsistencies
> > If we have structs, then functions should always operate on them. Since C99, compound literals makes life easier while passing temporaries to functions.
> 
> We should review what compilers and platforms are still supported. I'm still occasionally surprised that people are still using platforms that require C89 code. But I agree we should be consistent here, with an eye toward the most efficient and easy to use interfaces.

The C99 compound literals are for the user side, SDL does not need to have C99 support. This just means that if we take structures as argument the user can do:

```c
SDL_DoSomethingWithARectangle(&(const SDL_Rect){.w = 10 }));
```

> # Getting rid of dozens of buildsystems to unique one: CMake.
> I think we should bite the bullet and just pick CMake over autotools. Maintaining two separate configuration and build systems is too high a tax, and enough people (and platform maintainers) are using CMake that it seems like the right choice for SDL 3. That said, we should still keep the Visual Studio and Xcode projects for people who are more 
comfortable using the IDE's.

CMake generates proper Visual Studio and Xcode projects. Visual Studio even has a nice integration.

> # Making addons first class citizen and built from the same source tree with proper naming
> The problem here, and the reason why the SDL addons have lagged behind in terms of releases, is these depend on a lot of third party libraries that are time consuming to build and maintain in-tree. git submodules may make this easier though, so this is definitely worth reviewing. The ability to easily add modules to your project with CMake makes this much more viable in the future.

We **don't** incorporate third party libraries. That's generally a very bad thing to do because of ABI incompatibilities (and there are too much systems and architectures). This just means that we conditionally add support to a feature if present. For example, if user selects MP3 support, then SDL_mixer will be linked against the system mad/mpg321 library. Nothing more :)

**Comment by DanielGibson** [comment]
# Make SDL_main a header-only lib

Right now using SDL_main can be painful, especially with MinGW: There you have to link `-lmingw32 -lSDL2main -lSDL2` in that exact order (you can add other libs in between, but libmingw32 must be before libSDL2main, and libSDL2main must be before libSDL2) - especially getting libmingw32 there isn't trivial, see https://github.com/libsdl-org/SDL/blob/1bfefb5b4d5fbd9899fca460442240dad2bd4510/sdl2-config.cmake.in#L27-L41
I think this happens because main being in a static lib (instead of a source file that gets compiled as part of the build, or a .o/.obj) seems to confuse MinGW, which otherwise would automagically link libmingw32.

Another (also Windows-specific) problem is that .lib files generated by MinGW don't work with modern MSVC versions ("module unsafe for SAFESEH"). For the SDL2.lib only used for lining the DLL it's not that bad, because it can relatively easily be generated from SDL2.dll *(create a .def with mingw's `gendef SDL2.dll` and then use  `lib.exe /machine:x86 /def:SDL2.def /out:SDL2.lib` in the VS developer console to generate the lib)* - but that doesn't help for SDL2main.lib, which is a proper static lib and not just this stub-thing used on Windows to link DLLs.
So binary builds of SDL2 for Windows that have been built with MinGW aren't really compatible with MSVC, mostly because of SDL2main.

These problems would vanish if SDL2main was a header-only lib that you just `#include` in your source file that implements `int main(int argc, char** argv)` (and only there). It could still do its `#define main SDL_main` trick, and it could contain the platform-specific SDL_main code (or nothing at all for most of the platforms which don't use it). SDL.h wouldn't `#include <SDL_main.h>` anymore (maybe it shouldn't do that either way).
Like `SDL_config.h` it would be generated when SDL is built (or probably just copied from some dir in src/ to include/).

As a nice side-effect, it would make it clearer when it is used an when not:
Right now if you don't want to use it on e.g. Windows, you need to `#define SDL_MAIN_HANDLED` before `#include <SDL.h>` (or in your build process), which is not obvious - with SDL_main as a header-only lib, you'd just do nothing, and do `#include <SDL_main.h>` if you want to use it.
Also I think that right now if you have any variable (or function, enum member, ...) called `main` - in any file that just includes SDL.h, not just the one with `main()` in it! - and that `#define main SDL_main` is active on your platform, that variable will be renamed as well, which could lead to confusion..

Furthermore, it would make it easier to copy the SDL_main implementation into your own source and adjust it to your needs.

**Comment by slouken** [comment]
> # Make SDL_main a header-only lib

I like this idea. @icculus, what do you think? Is this something we can implement in 2.0.22?

**Comment by icculus** [comment]
I like the idea but would keep this in the SDL 3 bucket.

**Comment by slouken** [comment]
Another idea for SDL3:
* Remove all the complexity around removing API entry points. You can disable subsystems, but the API entry points will remain, with dummy implementations.

**Comment by slouken** [comment]
IMPORTANT: We will also need sdl20-compat from the very beginning.

**Comment by DanielGibson** [comment]
BTW, are there plans for a SDL2.1 before SDL3?
(If so, making SDL_main a header-only lib only requires very minor adjustments to applications using SDL2 - adding one `#include` in one file - so it IMO could be done in 2.1)

**Comment by slouken** [comment]
SDL3 is the next major ABI breaking revision - what we previously were calling SDL 2.1.

**Comment by icculus** [comment]
So I'm just catching up on this, so I wanted to throw in my current opinions.

I want to be clear that when suggesting big changes my instinct is to say no, and--as you can see from my "hard no" about GitHub--I don't make all the decisions and sometimes it turns out I'm proven wrong.

# Stricter coding, naming style

I agree that SDL2 is extremely inconsistent in this regard, and we should definitely clean that up for SDL3, but I do not think it has to be as rigid as SDL_SubsystemVerbSubject.

# Semantic versioning

SDL versioning can change, but semantic versioning is deeply overrated in my opinion.

We tend to never break even minor interfaces before we decide to change everything at once, and we do this less than once a decade in general. In the meantime, we add stuff all the time and constantly fix bugs and occasionally we bundle this up as production ready. We don't generally put out releases more than a few times a year.

Please do not reply with how the semver philosophy should be applied to this. It's just dogma, as far as I'm concerned.

# Remove useless platforms

Every time I try to get rid of a platform or backend, _someone_ pops up to say they still need it.

For various backends, this is largely harmless, because they can be ignored and even bitrot without a serious cost to the project.

But even things like Sony PSP and OS/2 have active users and maintainers. A lot of things would be _hard_ to delete, is all I'm saying.

Arts can definitely go, though.  :)

# Build systems

I would _love_ to drop to just CMake (or an equivalent, if there's a better option), but there is a value in telling people they can just double-click the Visual Studio project file and build. Even though CMake can generate vs project files, this adds friction for the average windows and Mac developer.

Between CMake and autotools though, CMake sucks less and we frequently have issues where we have to fix problems in one or the other, so dumping autotools is okay with me.

It might be interesting to find a way to automate this on Windows/Mac, so the user double clicks a small included .exe and it gets CMake installed if necessary and generates the project files.

# Use doxygen documentation to avoid outdated information on the wiki

This is done. Changes to Doxygen comments in the headers automatically generate Wiki changes now and vice-versa. It generates manpages now, too!

This doesn't cover anything but functions...structs and stuff are not bridged, but this can change.


# Making addons first class citizen and built from the same source tree with proper naming

100% no.

# Get rid of SDL_net

100% yes.

**Comment by icculus** [comment]
Here's an old Trello board about ideas for future improvements to SDL that got propped up at some point, I think to collect stuff that was mentioned around other places. Not all of it is still relevant (or was ever necessarily good)...

https://trello.com/b/sXcrlXCD/sdl-wishlist

...but it's probably worth sifting through. I've made it read-only to prevent the discussion from migrating to there.

**Comment by mkalte666** [comment]
I might have missed it, but for an abi breaking revision i would love to see a more consistent access to platform-dependent things. 

The main thing that made me think about this was this issue over here (which, i realize, is another shameless bump, sorry x.x): #3360  

The most use out of this would be in the rendering system - access window handles for video, texture and context access for the renderer.  

Where to draw draw the line though? 
While i never needed access to the stuff governing the event loop it might be easy to argue "hey, the rendering system got functions to pull out everthing from opengl to vulkan,  we want it too!", and i don't know if thats a path one should walk.

**Comment by DanielGibson** [comment]
I think the plan is to **remove** `SDL_GL_BindTexture()`, see #2124 (and IMHO it's the right thing to do).

HOWEVER, icculus is working on a 3D API (see https://discourse.libsdl.org/t/future-3d-api/33485/23), and if I recall correctly, there were plans to allow interaction between that and SDL_Render (I guess SDL_Render would use the new 3D API under the hood).

**Comment by mkalte666** [comment]
I created #3360 initially to think about an alternative to just removing stuff.
Anyway, i completely forgot about  "Letting people mix and mingle" point in the post on the list though, when writing my suggestion above. It does supersede everything i would ever want, so, ugh oops?

Don't mind me, have a great day :D

**Comment by DanielGibson** [comment]
BTW, regarding that earlier post from @markand

> The C99 compound literals are for the user side, SDL does not need to have C99 support. This just means that if we take structures as argument the user can do:
> 
> ```c
> SDL_DoSomethingWithARectangle(&(const SDL_Rect){.w = 10 }));
> ```

Designing an API for compound literals + designated initializers hurts its usability IMO, especially for C++:
* C++ only supports designated initializers (that `{.w=10, .h=20}` syntax) in C++20 and newer, which is much too new to demand, IMO
* Unlike C, when using them in C++ you must use the correct order of the initialized struct members (`{.h=20, .w=10}` is illegal for `struct SDL_Rect { int x; int y; int w; int h; };` because `.w` comes before `.h`). That makes using them more tedious.
* Your example takes the address of an r-value (I assume SDL_DoSomethingWithARectangle() takes a `const SDL_Rect*` as an argument). That seems to work in C, but is illegal in C++, see https://gcc.godbolt.org/z/7eY6rEv38

**Comment by markand** [comment]
> BTW, regarding that earlier post from @markand
> 
> > The C99 compound literals are for the user side, SDL does not need to have C99 support. This just means that if we take structures as argument the user can do:
> > ```c
> > SDL_DoSomethingWithARectangle(&(const SDL_Rect){.w = 10 }));
> > ```
> 
> Designing an API for compound literals + designated initializers hurts its usability IMO, especially for C++:

As I have mentioned, it just about adding consistency with what functions should take. some functions use several `int`s and some use `SDL_Rect`. That's it. How do you pass the structure is up to you.

**Comment by icculus** [comment]
Async rwops: #1374

**Comment by slime73** [comment]
High DPI consistency - most of these are assuming SDL will stick with DPI-scaled units for window position/size (which I personally prefer):
* float coordinates available for mouse positions (comments of #2119).
* APIs or fields to query the DPI scale or size-in-points vs size-in-pixels of a particular display mode for a monitor (comments of #2119). I'm coming from an Apple-centric perspective for this so I'm not sure if it makes sense across operating systems.
* APIs to query the active or recommended DPI scale of a display?
* Make Android behave the same as other operating systems with respect to SDL_WINDOW_ALLOW_HIGHDPI and DPI-scaled units (#1677).
* Clean up any inconsistencies between DPI scaling behaviour across operating systems - especially Windows vs the rest. I'm not sure where we'll be with high DPI on Windows by the time SDL 3 rolls around, but I hope Eric's work will be merged into SDL by then. (comments of #2119 again).

Some of these could be done before a theoretical SDL 3 because they don't all break or change APIs, but it's probably best to think of them all in concert for SDL 3 in any case.

**Comment by icculus** [comment]
System notification API: #1932

**Comment by icculus** [comment]
Drag and drop improvements: #5292

**Comment by icculus** [comment]
Don't clamp IME input to fit in an event structure: #5363

**Comment by Cacodemon345** [comment]
The WinRT backend can't be dropped until RetroArch abandons support for Xbox One and UWP. And I don't see them doing that. Microsoft is also yet to actually provide the Win32 APIs in Xbox Dev Mode, so it still needs to stick around until UWP is finally slated to go away completely from Xbox.

**Comment by icculus** [comment]
Reconsider multiple audio device opens: #5485

**Comment by flibitijibibo** [comment]
Throwing this out there because it's something I consistently ran into while working on Wayland, and is a consistent pain point for other backends like X: The SDL_Window API spec should be modified to make all changes **buffered**, with a CommitWindowChanges call to push everything at once.

This fixes the main problem we've been having lately, which is that the order of window calls can finally stop mattering and SDL can enforce consistent behavior without having to fight the OS as much. On top of being a pretty major bugfix (see the numerous issues for this, usually it's when modifying a window size after leaving fullscreen) it's also an optimization; Wayland for example has to commit the window surface _every time a call is made_, while in reality we shouldn't have to do it more than once (or twice if you count presentation) per frame. Right now compositors _hate_ us for doing it like this, and when you actually look at the interaction at a low level it's easy to see why.

This is similar to the old SetVideoMode without _forcing_ SetVideoMode - someone could make multiple CommitWindowChanges calls if they absolutely needed to, while still knowing the consequences of not optimizing the batch of changes.

**Comment by icculus** [comment]
Minor gripe for SDL3's consideration: SDL_CreateRenderer taking an index of the backend to use is a pain in the butt.

It either shouldn't do this at all (to match how video/audio work....they can be forced with a hint or otherwise you get what SDL decides is best for your system), or they should take a string instead of an index (so apps don't have to add a for-loop where they query each renderer and strcmp to see if it's "opengl", just to get the index number...NULL would mean "SDL decides for me.")

We can leave the ability to iterate available renderers (since we'll need it for sdl20-compat anyhow), this is just about simplifying the interface.

**Comment by icculus** [comment]
Some things that aren't prefixed by `SDL_` ...

- `RW_SEEK_*`
- `KMOD_*`
- `AUDIO_*`

...does anyone know of a tool that will parse a C header and tell you all the symbols exported from it (including functions, enums, preprocessor defines)? It might be worth pointing it at SDL.h and seeing what it reports.

**Comment by DanielGibson** [comment]
Kinda prefixed with SDL, but different to all others constant names: `SDLK_*`

**Comment by Iniquitatis** [comment]
> On the other other hand, renaming key constants would make porting to SDL3 painful for lots of software (even more so if the same code should be compatible with both SDL2 and SDL3, with a few `#ifdef`s), because both `SDLK_*` and `SDL_SCANCODE_*` are likely to be used in huge switch/case blocks or tables that contain at least dozens of key/scan-codes..

`SDLK_\w+` -> `SDL_KEY_\1`, then check the diff?

**Comment by markand** [comment]
Basically it's matter of the following search/replace in your code.

```
s,SDLK_,SDL_KEY_,g
```

**Comment by markand** [comment]
> Some things that aren't prefixed by `SDL_` ...
> 
>     * `RW_SEEK_*`
> 
>     * `KMOD_*`
> 
>     * `AUDIO_*`
> 
> 
> ...does anyone know of a tool that will parse a C header and tell you all the symbols exported from it (including functions, enums, preprocessor defines)? It might be worth pointing it at SDL.h and seeing what it reports.

I think doxygen and its XML output can give you that (with settings to export undocumented members to be sure).

**Comment by slouken** [comment]
I don't think we need to rename SDLK_*

**Comment by xlassmgreat** [comment]
Thread-safety? Thread-safe rendering is a different thing but the other parts of SDL too are for designed singly threaded programs. Every single "thread-safable" thing should be thread-safe. I don't think thread-safe rendering cannot be achieved anywhere in the near future but I think we should dabble with some experimental thread-safe rendering. Stuff in SDL_RenderThreadsafe*? I can imagine many situations in which programmers would like to throw in some thread-safe rendering in the middle of the program. Of course that would have its own disadvantages like, why even render in more than one thread but it's the decision of the programmer. As thread-safe APIs like Vulkan and Metal don't have a state, they should also be friendlier to integrate with.

**Comment by icculus** [comment]
The SDL GPU stuff will support multiple threads. While there will be a render backend that uses it, I don't think making the 2D render API itself thread safe is worth it.

**Comment by markand** [comment]
> I don't think we need to rename SDLK_*

Why not? I think every macro, enum constant, function, types should start with SDL_. But if we accept to start with just SDL than we can shorten some symbols (SDLK_, SDLKMOD_, etc).

**Comment by xlassmgreat** [comment]
> The SDL GPU stuff will support multiple threads. While there will be a render backend that uses it, I don't think making the 2D render API itself thread safe is worth it.

If there will be a render backend that is thread-safe, why not have the API thread-safe as well? Yes, there's not much point but why make it deliberately unsafe?

**Comment by icculus** [comment]
It isn't a question of throwing a few mutexes around; the underlying APIs aren't thread safe in some cases.

Also: you can already draw tens of thousands of 2D sprites at 60fps...how much value are you getting by adding threading to this?

**Comment by smcv** [comment]
Technical debt to clean up next time ABI is broken: #5626

**Comment by DanielGibson** [comment]
`SDL_DisplayMode::refresh_rate` shouldn't be an `int`, but either a float or two ints (fractional-style).
Apparently the real refreshrate often is 59.95Hz or sth like that - and to add insult to injury, it seems like Windows reports 59 then with the commonly used WinAPI (that SDL currently uses).

Also, no idea *at all* how exactly variable refreshrate displays work/are used, like what the refreshrate the OS returns is then, if it's only really variable when running in (exclusive?) fullscreen or something, etc - but SDL might needs ways to get information about this (like if it's available at all and min/max frequencies, I guess?).

And, for reference, the Windows (7+) API to get the proper refreshrate (not rounded down integer) is `QueryDisplayConfig()` -> `DISPLAYCONFIG_MODE_INFO` -> `DISPLAYCONFIG_TARGET_MODE` -> `DISPLAYCONFIG_VIDEO_SIGNAL_INFO` which has a fractional `vSyncFreq` member.

**Comment by slime73** [comment]
Some [APIs](https://github.com/libsdl-org/SDL/blob/main/include/SDL_system.h), defines, and public headers have "iphone" in their name, because iOS used to be called iPhoneOS. They could be renamed for SDL3.

**Comment by icculus** [comment]
Not sure I actually want this, but making note so it isn't lost: recursive mutexes va mutexes used for condition variables: https://github.com/libsdl-org/SDL/pull/5479

**Comment by smcv** [comment]
Related to #5711/#5712, it might be easier to get right if `SDL_GetWindowWMInfo()` either returned an opaque callee-allocated object with accessors, or used a simple incrementing integer that is increased every time a new field is added (`SDL_SYSWMINFO_VERSION`?) instead of SDL's user-facing version number.

The Linux kernel uses the equivalent of `sizeof(SDL_SysWMinfo)` as the "version number", but that requires having something like this:

```
typedef struct { ... } SDL_SysWMinfo1;
typedef struct { SDL_SysWMinfo1 parent; ... } SDL_SysWMinfo2;
typedef struct { SDL_SysWMinfo2 parent; ... } SDL_SysWMinfo3;
typedef SDL_SysWMinfo3 SDL_SysWMinfo;

/* caller */
SDL_SysWMinfo info = { 0 };
SDL_GetWindowWMInfo(window, sizeof(info), &info);

/* in SDL */
if (size >= sizeof(SDL_SysWMinfo2)) {
    info2 = (SDL_SysWMinfo2 *) info;
    info2->foo = 23;
    info2->bar = ":0";
}
```

**Comment by Jack-Ji** [comment]
Just pitch in my little thought. Since we have landed arbitrary triangle rendering (SDL_RenderGeometry), wouldn't it be great if SDL renderer can provide more 2D primitives drawing out of box (circle/ellipse/curve/quad/polygon etc)?

**Comment by flibitijibibo** [comment]
For the audio system: Add ChannelMask to SDL_AudioSpec. Wavs can technically include this, and platforms that support GetAudioDeviceSpec expose this as well.

**Comment by slouken** [comment]
Remove the parameter for SDL_PauseAudioDevice() and add SDL_PlayAudioDevice() to reduce confusion about how to start playing audio after opening the device.

**Comment by flibitijibibo** [comment]
One other audio device feature: StreamCategory. This is in IMMDevice as `ERole` and Pulse/PipeWire have similar enums.

**Comment by flibitijibibo** [comment]
Working on something adjacent to this, so I'm writing it here while I'm submerged in the muck:

The RWops and filesystem components could be merged into a new filesystem API, designed less around wrapping `fopen` and more on abstracting "containers" - that is, most platforms now have APIs centered on the concept of "title" containers and "storage" containers; title containers are the read-only filesystem for the application content, and storage containers are for save/config data created by the application at runtime. Storage containers are notable because you're actually meant to have _multiple_ containers, specifically for multiple user accounts. Right now we depend on the PC doing this via OS accounts, but this may not be applicable for platforms like Steam Deck which use a single OS account with multiple Steam accounts, and it definitely isn't applicable on consoles where each controller will have some kind of account associated with it (be it a full account or a "guest" account).

A fully redesigned filesystem API that exposes such OS APIs is important for a number of reasons:
- It would fix a ton of issues in one go, i.e. #1374 #1214 #3121
- It would allow for _significantly_ improved console support (right now it's either "have a single-player game on a platform that magically makes accounts vanish" or roll your own filesystem API from scratch, for every application)
- It would also allow for us to better support container APIs now found on PC. The notable examples are Xbox Live storage and SteamRemoteStorage, which apps currently implement totally separate from normal PC storage, assuming it's implemented at all (most of the time, something like Steam autocloud is used instead). If Steam ever exposes multi-account support (which I've [asked for](https://flibitijibibo.com/SteamMultiUser.txt) a number of times, and Deck may add pressure here), traditional filesystem APIs would not be able to handle this well (if at all). If Valve's [software patent](https://patft.uspto.gov/netacgi/nph-Parser?Sect1=PTO1&Sect2=HITOFF&d=PALL&p=1&u=%2Fnetahtml%2FPTO%2Fsrchnum.htm&r=1&f=G&l=50&s1=11123634.PN.&OS=PN/11123634&RS=PN/11123634) actually becomes a product beyond Fossilized shaders/media, this is where the title container idea would likely come in.

At least on my end, I'm familiar with this design because of XNA, where we have a storage namespace to reimplement in addition to a "TitleContainer":

https://github.com/FNA-XNA/FNA/tree/master/src/Storage
https://github.com/FNA-XNA/FNA/blob/master/src/TitleContainer.cs

All content reading is meant to be done from TitleContainer (a requirement for Xbox 360), and saves are done by selecting a storage device (typically based on the player index and an associated user account), opening the container, then closing the container when finished (which then flushes/commits the changes). It's not _perfect_ but it would probably be my starting point if I were writing this myself.

**Comment by icculus** [comment]
On a call this morning, the thought that some of our auxillary subsystems (logging, asserts, hints, etc) might be more complex than they need to be. No specific action items at the moment, but in SDL3 it might be worth thinking through how to trim these down to just what they need.

The end result might be "they are fine as-is," and nothing changes, but it's important to remember that anything _can_ change for SDL3.

**Comment by icculus** [comment]
Remove SDL_SetWindowBrightness and SDL_SetWindowGammaRamp, and associated functions.

**Comment by smcv** [comment]
Remove the equivalent of `sdl2.m4` (`AM_PATH_SDL2`), document `PKG_CHECK_MODULES([SDL3], [sdl3])` as the recommendation for Autotools users.

CMake users should continue to use `find_package(SDL3)` and `SDL3::SDL3` (or `SDL3::SDL3-static`), no change there except for replacing 2 with 3.

Remove the equivalent of `sdl2-config`, document `pkg-config sdl3` as the direct replacement (but Autotools users should prefer `PKG_CHECK_MODULES`, and artisanal hand-crafted makefiles/scripts should check `$PKG_CONFIG` and use that if set, falling back to `pkg-config` if unset).

**Comment by flibitijibibo** [comment]
Less a new feature and more fixing an existing one: We should change CreateWindowFrom to use something similar to SysWMInfo rather than an opaque pointer. Using a versioned struct will let us avoid doing things like [reading a pointer from an environment variable](https://github.com/libsdl-org/SDL/blob/main/src/video/windows/SDL_windowswindow.c#L611), which I'm still surprised we've gotten away with for so long. (Not as bad as .NET asking developers to set a _function pointer_ as an environment variable, but pretty darn close...)

**Comment by DanielGibson** [comment]
If there's still things like `HAVE_STDLIB_H` in the public headers, it should be renamed to `SDL_HAVE_STDLIB_H` or similar to avoid clashes with headers from other libs.
Otherwise people using two libs that use that kind of naming at the same time get compiler warnings about redefining those #defines - I think I have seen that with SDL2 and libjpeg in the past

**Comment by icculus** [comment]
It would actually be nice if the installed headers just work without any SDL_config.h, honestly.

Generally any app can build against SDL's headers just using the minimal config file, but we install the machine-specific SDL_config.h, even though this is only crucial for building SDL itself in modern times.

For SDL3, let's have one set of static headers that are installed for public use.

**Comment by icculus** [comment]
SDL_GetCPUCount() could probably stand to be more useful:

- Is a given core physical or logical (is this a hyperthread? Maybe I don't want to assign heavy-lifting to that).
- Is a given core low-power (which you see in newer Apple hardware, and I assume eventually other major platforms, too).

Also, a way to set thread affinity would be nice.

Take all these things together, and you can build a job system where you have threads waiting for jobs on specific physical cores, so they work at full efficiency and don't compete, and can optionally assign less critical jobs to queues on low-powered cores.

**Comment by DanielGibson** [comment]
> Is a given core low-power (which you see in newer Apple hardware, and I assume eventually other major platforms, too).

Current Intel CPUs (Alder Lake) also have different kind of cores (see for example https://ark.intel.com/content/www/us/en/ark/products/134591/intel-core-i712700-processor-25m-cache-up-to-4-90-ghz.html), so this might even be relevant on normal Desktop platforms (though I have no idea how much control over which core to use a non-privileged process has)

**Comment by icculus** [comment]
While I have wiki on the brain, I think we'll copy the current set of pages under `https://wiki.libsdl.org/SDL2/` and tweak the wiki bridge so that subdir will continue to update from a specific git branch's headers. `https://wiki.libsdl.org/PAGENAME` will be the SDL3 version, and if there's a matching page name under SDL2, we'll generate a header saying "the SDL2 version is _over here_".

The other option is just to manage this by hand. Functions that are gone will have a "***this function was in SDL2, but no longer exists in SDL3***" header on the page (the wiki bridge does not delete pages), and if the function changed in other ways, just have an "SDL2 Quirks" section in the text.

**Comment by aganm** [comment]
I do not know if the new renderer API already supports this but I have a suggestion: I would like for the new rendering API to support both AoS and SoA data layouts.

To give an example, the way `SDL_RenderGeometry` works is by taking a single pointer to an interleaved buffer, which is an Array of Structures (AoS). I would like if the equivalence for the new rendering API of `SDL_RenderGeometry` also had a way to take multiple pointers to distinct subbuffers of data instead, or Structure of Arrays (SoA).

If this is the current `SDL_RenderGeometry` function:

https://github.com/libsdl-org/SDL/blob/a7066c527aa18364088bdfed0d1c44603f53cb78/include/SDL_render.h#L1644-L1647

This is what I imagine this alternative `SDL_RenderGeometry` to look like: 

```C
typedef struct SDL_SubBuffers {
       float *x_values;
       float *y_values;
       float *z_values;
       float *s_values;
       float *t_values;
       float *r_values;
       float *g_values;
       float *b_values;
       float *a_values;
       uint8_t *r8_values;
       uint8_t *g8_values;
       uint8_t *b8_values;
       uint8_t *a8_values;
} SDL_SubBuffers;

extern DECLSPEC int SDLCALL SDL_RenderGeometrySub(SDL_Renderer *renderer,
                                               SDL_Texture *texture,
                                               const SDL_SubBuffers *buffers, int num_vertices,
                                               const int *indices, int num_indices);
```

An OpenGL backend could implement this with `glBufferSubData`, one call per sub buffer.

The motive for such an API is for when the engine or game code is layed out in a SoA fashion. I understand that the new rendering API is supposed to be closer to the metal, and SoA data layouts is something that modern graphics APIs support natively. This would allow to send SoA data as-is to the GPU, removing the friction between a data oriented engine and the SDL renderer API.

**Comment by icculus** [comment]
The new API sets up rendering pipelines with an array of these:

```c
typedef struct SDL_GpuVertexAttributeDescription
{
    SDL_GpuVertexFormat format;
    Uint32 offset;
    Uint32 stride;
    Uint32 index;
} SDL_GpuVertexAttributeDescription;
```

And then you set vertex buffers for each index in that array. You can use the same buffer for multiple things, so you can set these up as an array of structs.

**Comment by icculus** [comment]
(Or a struct of arrays.)

**Comment by icculus** [comment]
We should grep the sources for "2.1" (for comments like "we should change this in 2.1"). Also, a grep for FIXME probably makes some sense, too, although I'm not sure I would commit to fixing all the FIXMEs, 3.0 or otherwise.  :)

**Comment by icculus** [comment]
(Just thinking out loud, this isn't locked down at all yet.)

Big Audio changes:

- Remove the 1.2 API (no more SDL_OpenAudio and things that implicitly act on device id 1, there will just be the SDL2ish SDL_OpenAudioDevice). I liked the simplicity and backwards compatibility of the 1.2 functions, but it's not a big ask to keep a device handle around, and it removes the confusion about which API set to use.
- Make explicit that devices can only be opened once, however:
- Make it so you can feed the device multiple streams. In practice, this will be the usual audio callback or queue that you currently use, but you assign several of them to a device, and SDL will pull data from them as needed and handle the conversion and mixing. This way, each thing that wants a device (or wants to build a simple mixer) can just feed SDL whatever it has and SDL will make it work behind the scenes.

**Comment by slouken** [comment]
Also, when we flip over to SDL3, we should merge as many pull requests as we can before moving lots of code around.

**Comment by sulix** [comment]
If it's not too backwards-looking, I'd really like an ``SDL_BlitSurfaceEx()`` function (or change to ``SDL_BlitSurface()``) to add a flags field. In particular, it'd be nice to have these options (ignore the made-up-on-the-spot names):
* ``SDL_BLIT_FLIPHORZ``: Flip the surface horizontally while blitting. DirectDraw provides this, and it's a pain to have to open-code a blitter just to use it.
* ``SDL_BLIT_FLIPVERT``: Flip the surface vertically. More for consistancy than anything else.
* ``SDL_BLIT_IGNOREPALETTE``: When blitting from a paletted surface to another paletted surface, just copy the raw indices across, don't try to find the equivalent colours if the palettes don't match. Not only does this match DirectDraw's behaviour, it also fits really well with almost every paletted app, which tend to only want to set a palette for the final displayed surface, not for every intermediate one. Can be easily hacked around, but still…

So if adding more features to the old Surface API isn't out-of-the-question, those are the things which crop up _all-the-time_ when using the surface API to port older stuff.

**Comment by flibitijibibo** [comment]
It was the only way I could do Wayland work on UE/Unity as well as other titles I didn't have source access to, so yes, definitely a useful thing to have!

**Comment by icculus** [comment]
It drives me nuts that we have `SDL_VIDEODRIVER` and `SDL_AUDIODRIVER` and `SDL_RENDER_DRIVER` (extra underscore). Let's make these all match in SDL3.

**Comment by icculus** [comment]
Mentioned in #6535, for future consideration:

> However SDL_SysWMinfo probably should expose a pointer to contentView.layer.

**Comment by Kontrabant** [comment]
On my mind while working on some keyboard stuff: SDL currently ties keyboard modifier state for keys like shift, ctrl and alt to the idea that these keys are being physically pressed, which is not necessarily true in an era of virtual keyboards, accessibility shortcuts, and key remapping.  Modifier keys can be latched/locked and thus active even when their associated physical keys are up, and this state can be cancelled by pressing the key again, in which case the modifier can be off even though the associated physical key is currently down.  There are also concepts like 'shiftlock' (which capslock can be remapped to on *nix platforms), where all keys behave as though shift is pressed even though it's not physically down, and pressing shift in this state deactivates the modifier, so this is another case where the modifier state is inverse to the physical state of the associated keys.

My proposal for SDL 3 is that `SDL_GetModState()` should just return whether or not a specific modifier is active like it currently does for capslock and numlock, with no left/right designation or assumption that the modifier state equates to a specific key currently being up/down. If an application needs to know the physical state of a particular key, it should query the current state of the keyboard.

**Comment by slime73** [comment]
I think it'd be useful to evaluate whether specific SDL hints should be moved to function calls.

The list of hints is pretty large right now. Many of them make sense as-is for various reasons including letting users override certain behaviour, but some might be better suited to actual function APIs instead. For example maybe some of the mouse relative mode hints.

Similarly some of the hints are effectively deprecated and should be removed for SDL3.

**Comment by slime73** [comment]
Remove accelerometer joysticks on Android/iOS now that a real sensor API exists: https://github.com/libsdl-org/SDL/issues/3012

**Comment by sridenour** [comment]
Please, for the love of God, don't go making SDL_ttf, SDL_image, and SDL_mixer part of SDL. Sorry, but they have too many dependencies, and I don't particularly care for dependency hell.

SDL having more consistent naming is a good thing. Kinda tired of having to guess if a given function is `SDL_RenderDoThing()` or `SDL_DoThingRender()`, and having it always be `SDL_RenderDoThing()` etc. is certainly easier to find via autocomplete.

**Comment by icculus** [comment]
(This is brought on by #6581 finding out SDL_GetVersion now takes 16 milliseconds.)

Idea: the Hint subsystem should work like quake cvars.

Instead of a callback and a generic SDL_GetHint() function, you do something like this:

static SDL_bool support_foobars = SDL_TRUE;
SDL_RegisterHintBool("SDL_SUPPORT_FOOBARS", &support_foobars);

This adds this to a list or hash for later use, and sets a default if an environment variable is already set. When someone calls SDL_SetHint("SDL_SUPPORT_FOOBARS", "1"), it updates the variable directly. If a hint is set before it's registered, we'll keep track of it and RegisterHint will work out the bookkeeping like it does for environment variables. One can also mark a hint as locked (once we use it, we don't want the value to change)

Now when SDL needs to check a hint, it just looks at a static variable instead of calling a function, and you don't need a callback just to update one. The end result is that hint lookups are always _extremely_ fast and take less effort to maintain individually, as the complexity moves to a single common piece of code (in the set and register functions).

We can still keep an optional callback mechanism if there's work that needs to be done when a variable is set, but I'm more inclined to say there should be a simple "this has changed since the last time you used it" flag that code can check before using the value.

SDL_GetHint() can remain, in case a hint needs to be looked up by name, but internally, SDL can just look at variables internally.

**Comment by icculus** [comment]
SDL_DYNAMIC_API needs to become SDL3_DYNAMIC_API, so apps can override SDL2 with an sdl2-compat build and _also_ override SDL3.

(Failing that, sdl2-compat needs to see this is set and unset it before loading SDL3. But I'm also thinking of the guy that had SDL 1.2 plugins under sdl12-compat that wanted to also talk to pieces of SDL2 directly.)

**Comment by SteelT1** [comment]
There is `SDL_GetSystemRAM()` to get the total amount of RAM, but there is no function to get the *available* amount of  RAM, that was reported by the system. 

So something like `SDL_GetFreeSystemRAM()` which returns what's available in would be a good addition for SDL3. Additionally I think `SDL_GetSystemRAM()` itself should instead return in bytes, instead of MiB, as that would give developers more control of it's usage.

**Comment by rtrussell** [comment]
Everybody seems to be in favour of getting rid of **SDL_net** but as somebody who makes heavy use of it, across a wide range of platforms, I'd like to understand better what the replacement is expected to be, especially on Android and iOS.  If the idea is to use Posix socket functions everywhere that's fine, I have no objection, but are they straightforwardly available on mobile platforms?

The biggest networking issue for me, by far, is the absence of any cross-platform support for SSL, which is becoming ubiquitous and is increasingly making plain unencrypted socket functions useless.  It would be wonderful if SDL 3 could address this somehow.

**Comment by smcv** [comment]
> The biggest networking issue for me, by far, is the absence of any cross-platform support for SSL

I don't think that's really within the scope of SDL, and I'd recommend using one of the major TLS/SSL libraries: OpenSSL, GNUTLS, NSS, that sort of thing. Writing an abstraction layer across various platforms' TLS/SSL interfaces, if done by someone who is not a domain expert, seems like a recipe for security vulnerabilities.

**Comment by smcv** [comment]
> It drives me nuts that we have SDL_VIDEODRIVER and SDL_AUDIODRIVER and SDL_RENDER_DRIVER (extra underscore). Let's make these all match in SDL3.

Should they also be `SDL3_WHATEVER` so that SDL 1/2 and SDL 3 can be controlled independently?

**Comment by flibitijibibo** [comment]
> > It drives me nuts that we have SDL_VIDEODRIVER and SDL_AUDIODRIVER and SDL_RENDER_DRIVER (extra underscore). Let's make these all match in SDL3.
> 
> Should they also be `SDL3_WHATEVER` so that SDL 1/2 and SDL 3 can be controlled independently?

We got hit by this pretty hard with people trying to enable Wayland support, it would break for old 1.2 titles if set globally. I'd be in favor of adding the major version number to the hints.

**Comment by rtrussell** [comment]
> I don't think that's really within the scope of SDL, and I'd recommend using one of the major TLS/SSL libraries

I don't understand why it's less "within the scope of SDL" than anything else.  SDL is supposed to be an abstraction layer which hides differences between OSes behind a common API, including graphics, audio, file-handling, fonts etc.  I don't see why networking, including SSL, should be an exception, not least because I'm sure some multi-player games could benefit from it.

**Comment by DanielGibson** [comment]
Because
> Writing an abstraction layer across various platforms' TLS/SSL interfaces, if done by someone who is not a domain expert, seems like a recipe for security vulnerabilities

Furthermore, there are several open source cross-platform networking libraries for games, for example
http://enet.bespin.org/
https://github.com/ValveSoftware/GameNetworkingSockets
https://github.com/networkprotocol/yojimbo
https://github.com/networkprotocol/netcode
and probably more

Networking is completely orthogonal to SDLs features, so there's no reason to not use a non-SDL library for networking together with SDL in the same game

**Comment by rtrussell** [comment]
> Networking is completely orthogonal to SDLs features,

I personally don't see that it's any more "orthogonal" than file handling (indeed to me they seem conceptually very similar) but I guess I'm not going to get anywhere pursuing this request.  Pity.

**Comment by DanielGibson** [comment]
I mostly agree about the file handling (though keep in mind that through loading images or sounds this *does* interact with other features), but at least that adds at lot less complexity

**Comment by icculus** [comment]
The render info we give people before actually creating a renderer is largely bogus guesses, so let's remove it.

Also, let's change it so you give us a name ("direct3d", "opengl", etc) instead of an index, and NULL for "give me the best default" instead of the current -1.

**Comment by flibitijibibo** [comment]
Would narrowing down the Actions make it any faster...? Usually when I'm checking CI I just look at the subtasks and skip the ones that don't matter too much, and AFAIK it doesn't affect performance. Should we just split the most popular Windows/macOS/Linux build configs into their own CI files?

**Comment by nightmareci** [comment]
Since many modern platforms support sleeping threads with a delay precision better than 1 millisecond, it'd be nice if an option to delay at higher precision were provided. SDL_Delay could remain as-is, millisecond precision, using `Sleep` for Windows, with a higher-precision alternative, probably nanosecond precision per POSIX nanosleep. Higher precision delays, down to 100 nanosecond interval precision, can be done on Windows with [waitable timers](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-setwaitabletimer).

**Comment by smcv** [comment]
To avoid this issue becoming unmanageable, it might be best to limit its scope to the things that can *only* happen at the beginning of a cycle, basically meaning incompatible changes.

For instance, @nightmareci's suggestion of a more precise delay function isn't necessarily something that has to happen at the beginning of SDL 3 or not at all: `SDL_DelayNanoseconds` could equally well be added in 3.4 or 3.12 or whatever (or even in 2.28, if any more SDL 2.x "feature" releases happen).

The thing that *does* have to happen at the beginning of a cycle is backwards-compatibility breaks, like if we wanted to rename `SDL_Delay` to `SDL_DelayMilliseconds` to make the unit clearer. (I'm not saying that's necessarily desirable, just using it as an example of a backwards-compatibility break that @nightmareci's suggestion could conceivably cause.)

**Comment by nightmareci** [comment]
> To avoid this issue becoming unmanageable, it might be best to limit its scope to the things that can _only_ happen at the beginning of a cycle, basically meaning incompatible changes.
> 
> For instance, @nightmareci's suggestion of a more precise delay function isn't necessarily something that has to happen at the beginning of SDL 3 or not at all: `SDL_DelayNanoseconds` could equally well be added in 3.4 or 3.12 or whatever (or even in 2.28, if any more SDL 2.x "feature" releases happen).
> 
> The thing that _does_ have to happen at the beginning of a cycle is backwards-compatibility breaks, like if we wanted to rename `SDL_Delay` to `SDL_DelayMilliseconds` to make the unit clearer. (I'm not saying that's necessarily desirable, just using it as an example of a backwards-compatibility break that @nightmareci's suggestion could conceivably cause.)

Alright, I'll put it in as an issue, for SDL2 or SDL3.

**Comment by slime73** [comment]
Should the iOS demos be removed (https://github.com/libsdl-org/SDL/tree/main/Xcode-iOS/Demos)? Cross-platform SDL test programs already exist and I think the iOS demos have historically been less maintained than most other parts of SDL.

(I'm not sure if this was brought up already but I couldn't find it in a short search at least.)

**Comment by icculus** [comment]
- Move all the wiki pages from MediaWiki format to MarkDown. ghwikipp supports either format, but the headers are already in MarkDown, so we could eliminate conversion in wikiheaders.pl and also be less inhibited in what we do to satisfy that script. Also, MarkDown sucks less to write in.

**Comment by icculus** [comment]
> To avoid this issue becoming unmanageable

I'm going to be splitting this out into issues and closing this soon, fwiw. A wishlist made more sense when they were just wishes.  :)

**Comment by icculus** [comment]
> Would narrowing down the Actions make it any faster...? Usually when I'm checking CI I just look at the subtasks and skip the ones that don't matter too much, and AFAIK it doesn't affect performance. Should we just split the most popular Windows/macOS/Linux build configs into their own CI files?

More or less, that's what I'm proposing, but GitHub Actions also has a cron-style interface, so we can just blast out a "build everything you can think of" on a once-a-day or whatever basis that's divorced from the smaller set of checks we want to see immediately and in-line.

As for ignoring them: it runs in parallel to some degree, but it stops all currently-running builds if one of them fails, afaik, so I also want to avoid not knowing if Mac is working immediately because the PlayStation 2 port broke 100 commits ago.

**Comment by icculus** [comment]
> Should they also be SDL3_WHATEVER so that SDL 1/2 and SDL 3 can be controlled independently?

Módulo some name changes, you pretty much want the same thing between each version of SDL, and in the compat scenario, you only ever deal with one implementation...whereas with the dynamic API, you have two incompatible interfaces operating in parallel and might want to override them both.

**Comment by icculus** [comment]
> Everybody seems to be in favour of getting rid of SDL_net but as somebody who makes heavy use of it, across a wide range of platforms, I'd like to understand better what the replacement is expected to be, especially on Android and iOS. If the idea is to use Posix socket functions everywhere that's fine, I have no objection, but are they straightforwardly available on mobile platforms?

I'm going to write this here, and if it makes sense we can reference this in a new SDL_net issue, but let's not carry on discussion about this in this thread.

I will say, upfront, that I expect we are going to be _buried_ in SDL3 work, while still attempting to keep up maintenance on SDL2 while we do it, for the foreseeable future, so I just don't see an upgrade to SDL_net happening for a long time unless someone else steps up to do it.

To answer this question: yes, everything uses BSD Sockets now, or WinSock, which is 95% BSD Sockets and the rest can be protected by a small `#ifdef` block at the start of the code that conceals the differences.

The hairiest places are:
- Emscripten, which offers a BSD Sockets API but _webpages cannot open sockets to random internet addresses_, so it fakes it with WebSockets, which is a reliable bidirectional stream _like_ a TCP socket, but isn't TCP, so you can't just connect to port 25 with a WebSocket and talk SMTP to a mail server. But here, it's not something we'd be able to _abstract_ in SDL_net or anything else...you really have to design for it, if you can design for it at all. It's a mess. One _could_ make a game that uses WebSockets on all platforms, as they are built on top of TCP streams, in which case they can be used cleanly from Emscripten, and RTSP replacing UDP in roughly the same way...but these are complicated, and in both cases _probably_ require encryption to work at all. And now you're writing a very large and scary piece of code.
- Encryption, at any layer, is a pain in the butt, for lots of reasons.
- There's a lot of things that aren't real time games that don't want TCP or UDP...they want HTTP. libcurl is pretty complex but handles all sorts of wild (and yet common) scenarios, but it's a big dependency that can be hard to build...so it makes sense to push this off to various system APIs, since Windows, Mac, Android, etc, all offer this and handle SSL, too. But then you end up writing this five times, which is a scenario where SDL tends to be helpful.

**Comment by icculus** [comment]
Okay, I think I've split out most relevant things to their own issues, thus ending this megathread. If you don't see your thing mentioned in a separate issue, above, then I either missed it, thought it was already fixed, thought it was being handled by an existing issue, or made an executive decision to say no.  :)

But I can't really stop you in that last case, so you should always feel welcome to open a new issue yourself if you feel strongly about it!

This thread is being closed now, though, so don't dump any more ideas in here, or they risk being forgotten. Now that we're doing SDL3 for real, just go ahead and open a new issue.

Thanks everyone!
